12570 - Fix Variant Unit field is Out of Sync with the Pop-out

chahmedejaz opened the pull request:
#### What? Why?

- Closes #12570 
- We are creating a new variant object and passing it to the new variant row.
- This variant is empty and hence has no `unit_value` present in it, due to which `display_as` field is also empty and variant is unable to be saved.
- This PR creates and uses the helper to build this new variant object such that it has the `unit_value` of 1 x `variant_unit_scale`,
- When the variant `unit_value` is present, then it sets the placeholder value for the `display_as` field, which was empty

#### What should we test?
- The expected behavior mentioned in the issue should meet

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

dacook reviewed (commented):
I guess this solves it by ensuring there's value in the underlying field, but I don't see an answer as to why it was out of sync with what's displayed on the screen. 


Looking at `_variant_row.html.haml`, I can see that `unit_value_with_description` introduces a default of `1`,  when `unit_value` is blank. 

Your solution ensures unit_value is never blank, with a different default from the last variant. I think 1 is a better default (at least that's the current behaviour).

So I think that your helper should provide a default of 1, and we can remove the `1` from `_variant_row`.

dacook commented:
Oh, it might be event more complicated than that. The 

We want `variant_unit_scale` to inherit from the product too. This affects what `1` means. We want the `unit_value_with_description` to appear as "1", for whatever the `variant_unit_scale` is (whether 1, or 1000 for example).

Sorry, this system is really complicated, it took me ages to understand it, and I was too tired to try and document it. I've created a screenshot to help: https://github.com/openfoodfoundation/openfoodnetwork/wiki/Product-amounts-units#current-state

Also I have a diagram I drew to show the relationship between fields:
![Variant units](https://github.com/openfoodfoundation/openfoodnetwork/assets/4188088/4f3c9058-2645-4c24-8ec9-fbd199e2b0ed)

chahmedejaz commented:
Sure, thanks, David. I'll dig into this a bit more.

dacook commented:
Hi @chahmedejaz , just letting you know there are more conflicts on the way 😞 
Gaetan is currently working on [move-variant-unit-attributes-to-variant](https://github.com/rioug/openfoodnetwork/commits/move-variant-unit-attributes-to-variant/) (part of [Product Refactor]( https://github.com/openfoodfoundation/openfoodnetwork/issues/9069)).
His PR won't be ready until at least next week though.

chahmedejaz commented:
> Hi @chahmedejaz , just letting you know there are more conflicts on the way 😞 
> Gaetan is currently working on [move-variant-unit-attributes-to-variant](https://github.com/rioug/openfoodnetwork/commits/move-variant-unit-attributes-to-variant/) (part of [Product Refactor]( https://github.com/openfoodfoundation/openfoodnetwork/issues/9069)).
> His PR won't be ready until at least next week though.

Thanks for the heads up @dacook. I'll wrap this ASAP.

chahmedejaz commented:
@dacook - PR is ready for review now. Thanks

rioug reviewed (approved):
That looks good, thanks @chahmedejaz !
It's going to clash with the product refactor but I'll address that in due time.

dacook commented on app/views/admin/products_v3/_variant_row.html.haml:
```diff
@@ -21,7 +21,7 @@
     .field
       = f.label :display_as, t('admin.products_page.columns.display_as')
-      = f.text_field :display_as, placeholder: VariantUnits::OptionValueNamer.new(variant).name
+      = f.text_field :display_as, 'aria-label': t('admin.products_page.columns.display_as'), placeholder: VariantUnits::OptionValueNamer.new(variant).name
```
Above, `f.label :display_as` should produce a label with ID pointing to this text field. Therefore, the additional `aria-label` shouldn't be necessary. I think Capybara should be able to `find_field "Display unit as", placeholder: "1kg"` without any help.

Hmm, but the `find` in the spec isn't doing anything, it should be an expectation..

chahmedejaz replied:
Sorry about the aria-label. That's my bad.

> but the find in the spec isn't doing anything, it should be an expectation..

Ideally yes. I used the `find` method because it raises the exception `ElementNotFound` if it's unable to find the element otherwise It goes on.
But yes, expectation is better as it's more clear what the spec is doing. Thanks for rasing it. 👍

dacook commented on spec/system/admin/products_v3/create_spec.rb:
```diff
@@ -120,4 +132,12 @@ def visit_products_page_as_admin
     visit spree.admin_products_path
   end
+
+  def new_variant_button
+    find("button.secondary.condensed.naked.icon-plus")
+  end
```
In the previous spec, a different method, based on the button text is used. I think that would be better here, and it could be shared:
```
 page.find('button[aria-label="New variant"]', text: "New variant", visible: false)
```

dacook commented on spec/system/admin/products_v3/create_spec.rb:
```diff
@@ -45,6 +45,18 @@
+        expect(unit_button.text.strip).to eq('1kg')
+
+        unit_button.click
+        find('input[aria-label="Display unit as"][placeholder="1kg"]')
```
Shouldn't this be an expectation? Eg:
```suggestion
        expect(page).to have_field "Display unit as", placeholder: "1kg"
```

dacook reviewed (commented):
Yay, thanks for fixing this longrunning issue!

I think this is good, but I have a few questions where I think the code could be simpler.

chahmedejaz commented:
> Yay, thanks for fixing this longrunning issue!
> 
> I think this is good, but I have a few questions where I think the code could be simpler.

@dacook - I've addressed your comments in this commit here: https://github.com/openfoodfoundation/openfoodnetwork/pull/12602/commits/5b7fbc875ab4175e7a0e5434ad97108e12e3c399 Thanks.

dacook reviewed (approved):
Looks great!

drummer83 commented:
Hi @chahmedejaz,
Thanks for working on this fix.

## Before staging
I could not reproduce the exact original issue:
- After adding a new variant I was able to save without error (as long as any other information was changed).
- The unit value was always set in such way that it will be displayed as 1 g (or 1 L for Volume):
  - 0.001 for Weight (kg)
  - 1 for Weight (g)
  - 1000 for Weight (mg)
- The unit value however is only displayed after changing it (it returns to blank if closing the pop-out without changing the value).

## After staging
- The unit value is always set to 1.
- The unit value is immediately displayed after adding the new variant.

## Conclusion
I found no issues with your solution.
Ready to go! 🚀 Merging!! 🥳 
Thanks again!
